### PR TITLE
scarab: .NET 6 -> 8

### LINKED ISSUE #360336: Build failure: scarab depends on insecure dotnet-runtime-6.0.36
## Steps To Reproduce

1. build *scarab*

## Build log

<details>

<summary>Build Log</summary>

```
❯ git checkout 0ea5454f75c7027e3e1492369e40ed5e9294f094
❯ nix-build -A scarab
error:
       … while calling the 'derivationStrict' builtin
         at <nix/derivation-internal.nix>:34:12:
           33|
           34|   strict = derivationStrict drvAttrs;
             |            ^
           35|

       … while evaluating derivation 'scarab-2.5.0.0'
         whose name attribute is located at /home/victor/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:336:7

       … while evaluating attribute 'nativeBuildInputs' of derivation 'scarab-2.5.0.0'
         at /home/victor/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:380:7:
          379|       depsBuildBuild              = elemAt (elemAt dependencies 0) 0;
          380|       nativeBuildInputs           = elemAt (elemAt dependencies 0) 1;
             |       ^
          381|       depsBuildTarget             = elemAt (elemAt dependencies 0) 2;

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: Package ‘dotnet-runtime-6.0.36’ in /home/victor/nixpkgs/pkgs/development/compilers/dotnet/build-dotnet.nix:212 is marked as insecure, refusing to evaluate.


       Known issues:
        - Dotnet SDK 6.0.36 is EOL, please use 8.0 (LTS) or 9.0 (Current)

       You can install it anyway by allowing this package, using the
       following methods:

       a) To temporarily allow all insecure packages, you can use an environment
          variable for a single invocation of the nix tools:

            $ export NIXPKGS_ALLOW_INSECURE=1

          Note: When using `nix shell`, `nix build`, `nix develop`, etc with a flake,
                then pass `--impure` in order to allow use of environment variables.

       b) for `nixos-rebuild` you can add ‘dotnet-runtime-6.0.36’ to
          `nixpkgs.config.permittedInsecurePackages` in the configuration.nix,
          like so:

            {
              nixpkgs.config.permittedInsecurePackages = [
                "dotnet-runtime-6.0.36"
              ];
            }

       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
          ‘dotnet-runtime-6.0.36’ to `permittedInsecurePackages` in
          ~/.config/nixpkgs/config.nix, like so:

            {
              permittedInsecurePackages = [
                "dotnet-runtime-6.0.36"
              ];
            }
```

</details>

## Additional context

See also #360335.

## Metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.63, NixOS, 25.05 (Warbler), 25.05beta712559.4633a7c72337`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.10`
 - channels(root): `"home-manager, nixos, nixos-hardware, nixos-stable-24.05"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

## Notify maintainers

@huantianad

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by gepbird**
Closes https://github.com/NixOS/nixpkgs/issues/360336

Tracking: https://github.com/NixOS/nixpkgs/issues/326335

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by gepbird** [comment]
Related to failing nixfmt check: https://github.com/NixOS/nixpkgs/issues/358025, one of the proposals was to ignore it for generated files.

**Comment by gepbird** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 361251`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>scarab</li>
  </ul>
</details>

**Review by huantianad — APPROVED** [review]
Diff looks good to me, I probably should go test it with the game just to make sure, but I don't expect anything to go wrong.

Might be worth bothering upstream to make a new release with dotnet 8, might do that later.

**Comment by gepbird** [comment]
@huantianad 

> Diff looks good to me, I probably should go test it with the game just to make sure, but I don't expect anything to go wrong.

Proper testing would be very much appreciated, I only checked that the GUI opens.

> Might be worth bothering upstream to make a new release with dotnet 8, might do that later.

https://github.com/fifty-six/Scarab/pull/234

**Comment by huantianad** [comment]
Did a quick test: Fresh steam install of HK, Scarab detected it, and was able to install the modding API + a mod (Benchwarp). Game launched properly and mod was enabled. Disabling and uninstalling the mod also worked.

**Comment by gepbird** [comment]
cc @corngood @emilazy (https://github.com/NixOS/nixpkgs/pull/358533)

File: pkgs/by-name/sc/scarab/package.nix [hunk 1]
```diff
@@ -15,21 +15,20 @@ buildDotnetModule rec {
...
   testProjectFile = "Scarab.Tests/Scarab.Tests.csproj";
   executables = [ "Scarab" ];
 
-  preConfigureNuGet = ''
-    # This should really be in the upstream nuget.config
-    dotnet nuget add source https://api.nuget.org/v3/index.json \
-      -n nuget.org --configfile NuGet.Config
+  postPatch = ''
+    substituteInPlace Scarab/Scarab.csproj Scarab.Tests/Scarab.Tests.csproj \
+      --replace-fail 'net6.0' 'net8.0'
```
[the remark this answers is not in the document]
Comment (gepbird) [reply pkgs/by-name/sc/scarab/package.nix:31]: That's what I tried initially but it didn't work, any idea why?

<details><summary>Full diff</summary>

```diff
diff --git a/pkgs/by-name/sc/scarab/package.nix b/pkgs/by-name/sc/scarab/package.nix
index dac80b14f84f..f3f973ab977e 100644
--- a/pkgs/by-name/sc/scarab/package.nix
+++ b/pkgs/by-name/sc/scarab/package.nix
@@ -25,11 +25,7 @@ buildDotnetModule rec {
   projectFile = "Scarab/Scarab.csproj";
   testProjectFile = "Scarab.Tests/Scarab.Tests.csproj";
   executables = [ "Scarab" ];
-
-  postPatch = ''
-    substituteInPlace Scarab/Scarab.csproj Scarab.Tests/Scarab.Tests.csproj \
-      --replace-fail 'net6.0' 'net8.0'
-  '';
+  dotnetFlags = [ "-p:TargetFramework=net8.0" ];

   runtimeDeps = [
     bc
```
</details>

<details><summary>Build log</details>

```
Running phase: buildPhase
@nix { "action": "setPhase", "phase": "buildPhase" }
Executing dotnetBuildHook
  Scarab -> /build/source/Scarab/bin/Release/net8.0/linux-arm64/
Scarab.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:58.47
/nix/store/xdvhnm1rqgswk208ydl86i7gvac4q6cp-dotnet-sdk-8.0.404/s
hare/dotnet/sdk/8.0.404/Sdks/Microsoft.NET.Sdk/targets/Microsoft
.PackageDependencyResolution.targets(266,5): error NETSDK1005: A
ssets file '/build/source/Scarab/obj/project.assets.json' doesn'
t have a target for 'net6.0'. Ensure that restore has run and th
at you have included 'net6.0' in the TargetFrameworks for your p
roject. [/build/source/Scarab/Scarab.csproj]

Build FAILED.

/nix/store/xdvhnm1rqgswk208ydl86i7gvac4q6cp-dotnet-sdk-8.0.404/s
hare/dotnet/sdk/8.0.404/Sdks/Microsoft.NET.Sdk/targets/Microsoft
.PackageDependencyResolution.targets(266,5): error NETSDK1005: A
ssets file '/build/source/Scarab/obj/project.assets.json' doesn'
t have a target for 'net6.0'. Ensure that restore has run and th
at you have included 'net6.0' in the TargetFrameworks for your p
roject. [/build/source/Scarab/Scarab.csproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:07.14
```
</details>

[the remark this answers is not in the document]
Comment (corngood) [reply pkgs/by-name/sc/scarab/package.nix:31]: Hmm, I played with it a bit, but I don't really understand why that's broken.  I'll try to investigate more in the future.  Thanks for checking.

**Review by corngood — APPROVED** [review]

**Comment by corngood** [comment]
Unfortunately this has conflicts due to #355700.  @gepbird could you rebase?

**Comment by gepbird** [comment]
Thanks for the ping @corngood, done.

**Comment by corngood** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 361251`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>scarab</li>
  </ul>
</details>

**Pull request merged**